opencensus/exporters/trace/zipkin/internal/zipkin_exporter.cc

```diff
@@ -0,0 +1,361 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opencensus/exporters/trace/zipkin/zipkin_exporter.h"
+
+#include <arpa/inet.h>
+#include <ifaddrs.h>
+#include <iostream>
+
+#include <curl/curl.h>
+#include <rapidjson/prettywriter.h>
+#include <rapidjson/stringbuffer.h>
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "opencensus/trace/exporter/attribute_value.h"
+
+namespace opencensus {
+namespace exporters {
+namespace trace {
+
+namespace {
+
+constexpr char kZipkinLib[] = "zipkin/2.0";
+constexpr char ipv4_loopback[] = "127.0.0.1";
+constexpr char ipv6_loopback[] = "::1";
+
+std::string AttributeValueToString(
+    const ::opencensus::trace::exporter::AttributeValue &value) {
+  switch (value.type()) {
+    case ::opencensus::trace::AttributeValueRef::Type::kString:
+      return value.string_value();
+      break;
+    case ::opencensus::trace::AttributeValueRef::Type::kBool:
+      return value.bool_value() ? "true" : "false";
+      break;
+    case ::opencensus::trace::AttributeValueRef::Type::kInt:
+      return std::to_string(value.int_value());
+      break;
+  }
+  return "";
+}
+
+std::string SerializeAnnotation(
+    const ::opencensus::trace::exporter::Annotation &annotation) {
+  std::string annotation_str(annotation.description());
+  absl::StrAppend(&annotation_str, " (");
```

Comment:
not sure if others are doing this approach to serializing annotation attributes. At least if there are no attribute, I'd leave off the " ()" suffix as in zipkin annotations are equals search

Reply:
Done.